File: ios/LivtetJigsaw/Sources/LivtetJigsaw/Model/Connector.swift
import CoreFoundation

public typealias ConnectionRequirement = (Piece, Piece) -> Bool
public let noConnectionRequirements: ConnectionRequirement = { _, _ in true }

public protocol Connector: AnyObject {
    func closeTo(_ a: Piece, _ b: Piece, proximity: CGFloat) -> Bool
    func match(_ a: Piece, _ b: Piece) -> Bool
    func canConnectWith(_ a: Piece, _ b: Piece, proximity: CGFloat) -> Bool
    func connectWith(_ a: Piece, _ b: Piece, proximity: CGFloat, back: Bool)
    func attract(_ a: Piece, _ b: Piece, back: Bool)
}

public final class HorizontalConnector: Connector {
    private var requirement: ConnectionRequirement = noConnectionRequirements

    public init() {}

    public func closeTo(_ a: Piece, _ b: Piece, proximity: CGFloat) -> Bool {
        let forward = Vector.isClose(a.rightAnchor.asVector(), b.leftAnchor.asVector(), tolerance: proximity)
        let reverse = Vector.isClose(b.rightAnchor.asVector(), a.leftAnchor.asVector(), tolerance: proximity)
        return forward || reverse
    }

    public func match(_ a: Piece, _ b: Piece) -> Bool {
        (a.right.complement == b.left) || (a.left.complement == b.right)
    }

    public func canConnectWith(_ a: Piece, _ b: Piece, proximity: CGFloat) -> Bool {
        closeTo(a, b, proximity: proximity) && (match(a, b) || a.connected || b.connected) && requirement(a, b)
    }

    public func connectWith(_ a: Piece, _ b: Piece, proximity: CGFloat, back: Bool) {
        if a.right.complement == b.left {
            a.rightConnection = b
            b.leftConnection = a
            a.fireConnect(b)
            b.fireConnect(a)
        }
        if back && a.left.complement == b.right {
            a.leftConnection = b
            b.rightConnection = a
        }
    }

    public func attract(_ a: Piece, _ b: Piece, back: Bool) {
        if a.right.complement == b.left {
            b.recenterAround(a.rightAnchor)
            connectWith(a, b, proximity: 0, back: back)
        }
    }

    public func attachRequirement(_ req: @escaping ConnectionRequirement) {
        requirement = req
    }
}

public final class VerticalConnector: Connector {
    private var requirement: ConnectionRequirement = noConnectionRequirements

    public init() {}

    public func closeTo(_ a: Piece, _ b: Piece, proximity: CGFloat) -> Bool {
        let forward = Vector.isClose(a.downAnchor.asVector(), b.upAnchor.asVector(), tolerance: proximity)
        let reverse = Vector.isClose(b.downAnchor.asVector(), a.upAnchor.asVector(), tolerance: proximity)
        return forward || reverse
    }

    public func match(_ a: Piece, _ b: Piece) -> Bool {
        (a.down.complement == b.up) || (a.up.complement == b.down)
    }

    public func canConnectWith(_ a: Piece, _ b: Piece, proximity: CGFloat) -> Bool {
        closeTo(a, b, proximity: proximity) && (match(a, b) || a.connected || b.connected) && requirement(a, b)
    }

    public func connectWith(_ a: Piece, _ b: Piece, proximity: CGFloat, back: Bool) {
        if a.down.complement == b.up {
            a.downConnection = b
            b.upConnection = a
            a.fireConnect(b)
            b.fireConnect(a)
        }
        if back && a.up.complement == b.down {
            a.upConnection = b
            b.downConnection = a
        }
    }

    public func attract(_ a: Piece, _ b: Piece, back: Bool) {
        if a.down.complement == b.up {
            b.recenterAround(a.downAnchor)
            connectWith(a, b, proximity: 0, back: back)
        }
    }

    public func attachRequirement(_ req: @escaping ConnectionRequirement) {
        requirement = req
    }
}

File: android/jigsaw/src/main/kotlin/net/olamaelcu/livtet/jigsaw/model/Connector.kt
package net.olamaelcu.livtet.jigsaw.model

typealias ConnectionRequirement = (Piece, Piece) -> Boolean
val noConnectionRequirements: ConnectionRequirement = { _, _ -> true }

enum class ConnectorKind { Horizontal, Vertical }

class Connector internal constructor(kind: ConnectorKind) {
    val axis = if (kind == ConnectorKind.Horizontal) 'x' else 'y'
    val forward = if (kind == ConnectorKind.Horizontal) "right" else "down"
    val backward = if (kind == ConnectorKind.Horizontal) "left" else "up"
    val forwardAnchor get() = "${forward}Anchor"
    val backwardAnchor get() = "${backward}Anchor"
    val forwardConnection get() = "${forward}Connection"
    val backwardConnection get() = "${backward}Connection"
    var requirement: ConnectionRequirement = noConnectionRequirements

    fun attract(one: Piece, other: Piece, back: Boolean = false) {
        val (iron, magnet) = if (back) one to other else other to one
        val mc = magnet.centralAnchor ?: return
        val ic = iron.centralAnchor ?: return
        val ma: Anchor
        val ia: Anchor
        if ((axis == 'x' && mc.x > ic.x) || (axis == 'y' && mc.y > ic.y)) {
            ma = magnet.getAnchor(backwardAnchor) ?: return
            ia = iron.getAnchor(forwardAnchor) ?: return
        } else {
            ma = magnet.getAnchor(forwardAnchor) ?: return
            ia = iron.getAnchor(backwardAnchor) ?: return
        }
        val (dx, dy) = ma.diff(ia)
        iron.push(dx, dy)
    }

    fun openMovement(one: Piece, delta: Float) =
        (delta > 0 && one.getConnection(forwardConnection) == null) ||
        (delta < 0 && one.getConnection(backwardConnection) == null) || delta == 0f

    fun canConnectWith(one: Piece, other: Piece, proximity: Float) =
        closeTo(one, other, proximity) && match(one, other) && requirement(one, other)

    fun closeTo(one: Piece, other: Piece, proximity: Float): Boolean {
        val f = one.getAnchor(forwardAnchor) ?: return false
        val b = other.getAnchor(backwardAnchor) ?: return false
        return f.closeTo(b, proximity)
    }

    fun match(one: Piece, other: Piece) =
        one.getInsert(forward).match(other.getInsert(backward))

    fun connectWith(one: Piece, other: Piece, proximity: Float, back: Boolean = false) {
        require(canConnectWith(one, other, proximity)) { "can not connect $forward!" }
        if (one.getConnection(forwardConnection) !== other) {
            attract(other, one, back)
            one.setConnection(forwardConnection, other)
            other.setConnection(backwardConnection, one)
            one.fireConnect(other)
        }
    }
}
